### PR TITLE
refactor packageuploadservice

### PR DESCRIPTION
**Pull request by lyndaidaii**
Summary of the changes (in less than 80 characters):

separate validation before uploading packages

additional work need to be done in terms of separating validation in validation service

Addresses NuGet/Engineering#2342

### CODE REVIEW

File: tests/NuGetGallery.Facts/NuGetGallery.Facts.csproj [hunk 1]
```diff
@@ -375,11 +376,31 @@
   <ItemGroup>
     <WCFMetadata Include="Connected Services\" />
   </ItemGroup>
+  <ItemGroup />
```
Comment (lyndaidaii) [inline tests/NuGetGallery.Facts/NuGetGallery.Facts.csproj:379]: delete those

File: src/NuGetGallery/Services/IPackageMetadataValidationService.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+using System.Threading.Tasks;
+using NuGet.Packaging;
+using NuGet.Services.Entities;
+using NuGetGallery.Packaging;
+
+namespace NuGetGallery
+{
+    public interface IPackageMetadataValidationService
+    {
+        Task<PackageValidationResult> ValidateMetadataBeforeUploadAsyn(
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/IPackageMetadataValidationService.cs:13]: Please add a comment on this method

Comment (loic-sharma) [inline src/NuGetGallery/Services/IPackageMetadataValidationService.cs:13]:
```suggestion
        Task<PackageValidationResult> ValidateMetadataBeforeUploadAsync(
```

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs [hunk 3]
```diff
@@ -0,0 +1,678 @@
...
+            if (result != null)
+            {
+                _telemetryService.TrackLicenseValidationFailure();
+                return result;
+            }
+
+            result = await CheckIconMetadataAsync(nuGetPackage, warnings, currentUser);
+            if (result != null)
+            {
+                //_telemetryService.TrackIconValidationFailure();
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/PackageMetadataValidationService.cs:140]: /cc @agr Is this expected? Should there be an issue tracking this telemetry or should we remove it now?

**Comment by loic-sharma** [comment]
Here are some diffs to help you review this PR:

* https://www.diffchecker.com/apJ4drcs

File: src/NuGetGallery/Services/PackageUploadService.cs [hunk 4]
```diff
@@ -9,55 +9,16 @@
...
-        /// </remarks>
-        private const long MaxAllowedLicenseLengthForUploading = 1024 * 1024; // 1 MB
-        private const long MaxAllowedIconLengthForUploading = 1024 * 1024; // 1 MB
-        private const int MaxAllowedLicenseNodeValueLength = 500;
-        private const string LicenseNodeName = "license";
-        private const string IconNodeName = "icon";
-        private const string AllowedLicenseVersion = "1.0.0";
-        private const string Unlicensed = "UNLICENSED";
 
         private readonly IPackageService _packageService;
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/PackageUploadService.cs:23]: Nit: there's an empty line above this field

File: src/NuGetGallery/Services/PackageUploadService.cs [hunk 5]
```diff
@@ -785,6 +122,40 @@ private PackageValidationResult CheckRepositoryMetadata(PackageMetadata packageM
...
+                return result;
+            }
+
+            if (isNewPackageRegistration && _typosquattingService.IsUploadedPackageIdTyposquatting(package.Id, owner, out List<string> typosquattingCheckCollisionIds))
+            {
+                return PackageValidationResult.Invalid(string.Format(Strings.TyposquattingCheckFails, string.Join(",", typosquattingCheckCollisionIds)));
+            }
+
+            return PackageValidationResult.Accepted();
+        }
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/PackageUploadService.cs:158]: Nit: this is missing an empty line below this line

File: src/NuGetGallery/Services/PackageUploadService.cs [hunk 6]
```diff
@@ -93,641 +51,20 @@ public class PackageUploadService : IPackageUploadService
...
-
-            return PackageValidationResult.Accepted();
-        }
-
-        private async Task<PackageValidationResult> ValidateSignatureFilePresenceAsync(
-            PackageRegistration packageRegistration,
-            PackageArchiveReader nugetPackage,
-            User owner,
-            User currentUser)
+        public async Task<PackageValidationResult> ValidateSignatureFilePresenceAsync(
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/PackageUploadService.cs:63]: This method used to be below `ValidateAfterGeneratePackageAsync` - could we move it back down to reduce the diff? Also, the `ValidateBeforeGeneratePackageAsync` and `ValidateAfterGeneratePackageAsync` methods are the logical entry points, so having these methods be first in this class improves readability.

Comment (loic-sharma) [reply src/NuGetGallery/Services/PackageUploadService.cs:63]: Also, can we move this method to the `PackageMetadataValidationService`? If not, consider adding a comment explaining this

File: src/NuGetGallery/Services/PackageUploadService.cs [hunk 7]
```diff
@@ -785,6 +122,40 @@ private PackageValidationResult CheckRepositoryMetadata(PackageMetadata packageM
...
+                currentUser);
+            if (result != null)
+            {
+                return result;
+            }
+
+            if (isNewPackageRegistration && _typosquattingService.IsUploadedPackageIdTyposquatting(package.Id, owner, out List<string> typosquattingCheckCollisionIds))
+            {
+                return PackageValidationResult.Invalid(string.Format(Strings.TyposquattingCheckFails, string.Join(",", typosquattingCheckCollisionIds)));
+            }
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/PackageUploadService.cs:155]: Should we add a `IPackageMetadataValidationService.ValidateMetadataAfterUploadAsync` method that validates the signature and checks for typosquatting?

File: src/NuGetGallery/Services/IPackageMetadataValidationService.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        /// <summary>
+        /// Validate the provided package archive reader before generating package
+        /// This is useful for finding errors or warnings that should be caught before the user verifies their
+        /// UI package upload.
+        /// </summary>
+        /// <param name="nuGetPackage">The package archive reader.</param>
+        /// <param name="packageMetadata">The package metadata.</param>
+        /// <param name="currentUser">The user who pushed the package.</param>
+        /// <returns>The package validation result.</returns>
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/IPackageMetadataValidationService.cs:21]: Nice! These comments are great 👍

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs [hunk 9]
```diff
@@ -345,6 +345,11 @@ protected override void Load(ContainerBuilder builder)
                 .As<ISymbolPackageService>()
                 .InstancePerLifetimeScope();
 
+            builder.RegisterType<PackageMetadataValidationService>()
+                .AsSelf()
```
Comment (agr) [inline src/NuGetGallery/App_Start/DefaultDependenciesModule.cs:349]: Is there a need to add `AsSelf()` here?

File: src/NuGetGallery/Services/PackageMetadataValidationService.cs (hunk 3 above)
Comment (agr) [reply src/NuGetGallery/Services/PackageMetadataValidationService.cs:143]: I suppose it should be uncommented, but that method is not implemented in telemetry service. I'll file a bug.

Comment (agr) [reply src/NuGetGallery/Services/PackageMetadataValidationService.cs:143]: https://github.com/NuGet/Engineering/issues/3280

**Pull request merged**